### PR TITLE
add v1/v2 `TransactionInfo` and related types

### PR DESCRIPTION
**Pull request by chfanghr**
The following types are added in this pr:

* `v1::assoc_map::AssocMap`, also exported under `v2`.
* `v1::tuple::Tuple`, also exported under `v2`
* `v1::transaction::{DCert, ScriptPurpose, TransactionInfo, ScriptContext}`. `DCert` and `ScriptPurpose` are exported under `v2::transaction` as well.
* `v2::transaction::{TransactionInfo, ScriptContext}`

`IsPlutusData` for types mentioned are implemented; plutus data round-trip tests are implemented

TODO:

- [x] doc
- [x] json round-trip tests

### CODE REVIEW

**Review by szg251 — CHANGES_REQUESTED** [review]
Solid work.
Other than the thing I mentioned yesterday (docs and JSON instances) I commented on one small thing.

File: src/v1/transaction.rs [hunk 1]
```diff
@@ -14,6 +15,12 @@ use num_bigint::BigInt;
 #[cfg(feature = "serde")]
 use serde::{Deserialize, Serialize};
 
+use super::address::StakingCredential;
+use super::crypto::PaymentPubKeyHash;
+use super::datum::Datum;
+use super::tuple::Tuple;
+use super::value::CurrencySymbol;
+
```
Comment (szg251) [inline src/v1/transaction.rs:23]: [nit] It kind of feels odd that some modules are imported as `crate::v1::xyz` while some are `super::xyz`, could you change the all of the former to use super as well? Also, could you remove the empty line, to let the formatter auto-sort all imports?

File: src/v2/transaction.rs [hunk 2]
```diff
@@ -14,6 +18,14 @@ use num_bigint::BigInt;
 #[cfg(feature = "serde")]
 use serde::{Deserialize, Serialize};
 
+use super::{
```
Comment (szg251) [inline src/v2/transaction.rs:21]: The same applies here, could you move everything from `crate::v2` to `super`, and remove the empty line

File: src/v1/transaction.rs (hunk 1 above)
Comment (chfanghr) [reply src/v1/transaction.rs]: My editor did auto-import for me so I really don't know. Also, it seems that `cargo fmt` cannot combine imports, which is a bit inconvenient.

Comment (chfanghr) [reply src/v1/transaction.rs]: Anyway, I have replaced all the `crate::v1/v2::something` with `super::something`, but import sorting doesn't seem to work whenever there are `#[cfg(feature = "something")]` lines among the `use` lines.

File: src/v1/tuple.rs
Comment (szg251) [inline src/v1/tuple.rs:1]: I just realised, that there's already an implementation for Tuples here https://github.com/mlabs-haskell/plutus-ledger-api-rust/blob/master/src/plutus_data.rs#L521
(I don't think there's a difference, but I could be wrong)

**Review by szg251 — APPROVED** [review]
Looks good!

**Pull request merged**